File: Month 1- Java, Python, SQL Fundamentals/Java Fundamentals/LibraryBook_OOP_Practice/src/main/java/org/example/Book.java
package org.example;

public class Book {

    private String title, author, genre;
    private int publishYear, timesBorrowed;
    private boolean isBorrowed;

    //CONSTRUCTOR
    public Book(String title, String author, String genre, int publishYear) {

        this.author = author;
        this.title = title;
        this.publishYear = publishYear;
        this.genre = genre;

        timesBorrowed = 0;
        isBorrowed = false;
    }// End of constructor

    //end of method 1
    public void borrowBook() {

        if (isBorrowed) {
            System.out.println("Book is already borrowed.");
            return;
        } else {
            System.out.println("Book borrowed successfully.");
            isBorrowed = true;
            timesBorrowed++;//timesBorrowed = timesBorrowed + 1, THIS IS THE ALTERNATE VERSION
        }// end of else
    }//end of method

    //end of method 2
    public void returnBook() {

        if (!isBorrowed) {
            System.out.println("This book is already in the library.");
        } else {
            System.out.println("Book returned successfully.");
            isBorrowed = false;
        }//end of else

    }//end of method 2

    //method 3
    public void readBook() {

        if (isBorrowed) {
            System.out.println("Cannot read.\n" +
                    "Book has been borrowed.");

        } else {
            System.out.println("Reading Clean Code...");
        }// end of else
    }// end of method 3

    //return method
    public String getBookInfo() {

        return "Book Details: \n" +
                "Title: " + title +
                "\nAuthor: " + author +
                "\nYear Published: " + publishYear +
                "\nStatus: " + isBorrowed +
                "\nTimes Borrowed: " + timesBorrowed +
                "\n";
    }// END OF RETURN


}//END OF CLASS

File: Java-Python-SQL-Fundamentals/Python-Fundamentals/Book_management_system/Book.py
class Book:

    def __init__(self, title, author,genre, year_released):
        self.title = title
        self.author = author
        self.genre = genre
        self.year_released = year_released
        self.times_borrowed= 0
        self.is_borrowed= False
        self.available=True

    def borrow_book(self):

        if self.is_borrowed:
            print('Book already borrowed')
            return

        print('Book borrowed')
        self.is_borrowed = True
        self.times_borrowed+=1
        self.available=False

    def return_book(self):

        if not self.is_borrowed:
            print('Book is already in the library')
            return

        print('Book returned')
        self.is_borrowed = False
        self.available =True

    def get_book_info(self) -> str:

        return ( "Book details: \n"
               f'Title: {self.title} \n'
               f'Author: {self.author} \n'
               f'Genre: {self.genre}\n'
               f'Year Published: {self.year_released}\n'
               f'Book Available: {self.available} \n'
               f'Times Borrowed: {self.times_borrowed}\n')

book_1 = Book("ACE", "Ace Ventura", "Documentary", 2009)
book_2 = Book("Spies", "Larry Vena", "Action", 2023)
book_3 = Book("Jigga Maroo", "lex lura", "Sci-FI", 2014)

print(book_1.get_book_info())
print(book_2.get_book_info())
print(book_3.get_book_info())


book_1.borrow_book()
print()

book_1.return_book()
book_1.return_book()
print()

print(book_1.get_book_info())












